File: pdl/infra/php/src/Io/Pdl/Infrastructure/Data/Store/MemoryStore.php
<?php

namespace Io\Pdl\Infrastructure\Data\Store;

use Io\Pdl\Infrastructure\Data\DBStore;
use Io\Pdl\Infrastructure\Data\Operator;
use RuntimeException;

final class MemoryStore implements DBStore
{
    /** @var array<string, array<int, array<string, mixed>>> */
    private array $tables = [];

    public function insert(string $table, string $primaryKey, array $values): array
    {
        $tableData = &$this->tables[$table];
        if ($tableData === null) {
            $tableData = [];
        }
        if (!array_key_exists($primaryKey, $values) || $values[$primaryKey] === null) {
            $values[$primaryKey] = count($tableData) + 1;
        }
        $tableData[] = $values;
        return [$primaryKey => $values[$primaryKey]];
    }

    public function update(string $table, string $primaryKey, array $values): void
    {
        foreach ($this->tables[$table] ?? [] as &$row) {
            if (($row[$primaryKey] ?? null) === ($values[$primaryKey] ?? null)) {
                $row = array_replace($row, $values);
                return;
            }
        }
        throw new RuntimeException('MemoryStore: row not found for update');
    }

    public function delete(string $table, string $primaryKey, array $values): void
    {
        foreach ($this->tables[$table] ?? [] as $index => $row) {
            if (($row[$primaryKey] ?? null) === ($values[$primaryKey] ?? null)) {
                unset($this->tables[$table][$index]);
                return;
            }
        }
    }

    public function select(
        string $table,
        array $filters,
        array $projection,
        array $orderings,
        ?int $limit,
        ?int $offset
    ): array {
        $rows = array_values($this->tables[$table] ?? []);
        foreach ($filters as [$column, $operator, $value]) {
            $rows = array_values(array_filter($rows, static function (array $row) use ($column, $operator, $value): bool {
                $current = $row[$column] ?? null;
                return match ($operator) {
                    Operator::EQ => $current === $value,
                    Operator::NEQ => $current !== $value,
                    Operator::IN => is_array($value) && in_array($current, $value, true),
                    default => false,
                };
            }));
        }
        if (!empty($orderings)) {
            usort($rows, static function (array $left, array $right) use ($orderings): int {
                foreach ($orderings as [$column, $direction]) {
                    $comparison = ($left[$column] ?? null) <=> ($right[$column] ?? null);
                    if ($comparison !== 0) {
                        return $direction === 'DESC' ? -$comparison : $comparison;
                    }
                }
                return 0;
            });
        }
        if ($offset !== null) {
            $rows = array_slice($rows, $offset);
        }
        if ($limit !== null) {
            $rows = array_slice($rows, 0, $limit);
        }
        if (!empty($projection)) {
            $rows = array_map(static function (array $row) use ($projection): array {
                $projected = [];
                foreach ($projection as $column) {
                    $projected[$column] = $row[$column] ?? null;
                }
                return $projected;
            }, $rows);
        }
        return $rows;
    }
}

File: pdl/infra/csharp/Io/Pdl/Infrastructure/Data/Store/MemoryStore.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Io.Pdl.Infrastructure.Data.Store;

using Io.Pdl.Infrastructure.Data;

public sealed class MemoryStore : IDbStore
{
    private readonly Dictionary<string, List<Dictionary<string, object?>>> _tables = new();

    public IDictionary<string, object?> Insert(string table, string primaryKey, IDictionary<string, object?> values)
    {
        if (!_tables.TryGetValue(table, out var rows))
        {
            rows = new List<Dictionary<string, object?>>();
            _tables[table] = rows;
        }

        var row = new Dictionary<string, object?>(values);
        if (!row.TryGetValue(primaryKey, out var keyValue) || keyValue is null)
        {
            keyValue = rows.Count + 1;
            row[primaryKey] = keyValue;
        }

        rows.Add(row);
        return new Dictionary<string, object?> { [primaryKey] = keyValue };
    }

    public void Update(string table, string primaryKey, IDictionary<string, object?> values)
    {
        if (!_tables.TryGetValue(table, out var rows))
        {
            throw new InvalidOperationException("MemoryStore: table not found");
        }

        foreach (var row in rows)
        {
            if (Equals(row.GetValueOrDefault(primaryKey), values.GetValueOrDefault(primaryKey)))
            {
                foreach (var pair in values)
                {
                    row[pair.Key] = pair.Value;
                }
                return;
            }
        }

        throw new InvalidOperationException("MemoryStore: row not found");
    }

    public void Delete(string table, string primaryKey, IDictionary<string, object?> values)
    {
        if (!_tables.TryGetValue(table, out var rows))
        {
            return;
        }

        rows.RemoveAll(row => Equals(row.GetValueOrDefault(primaryKey), values.GetValueOrDefault(primaryKey)));
    }

    public IReadOnlyList<IDictionary<string, object?>> Select(
        string table,
        IReadOnlyList<Filter> filters,
        IReadOnlyList<string> projection,
        IReadOnlyList<OrderClause> orderings,
        int? limit,
        int? offset)
    {
        var rows = _tables.TryGetValue(table, out var data)
            ? data.Select(row => new Dictionary<string, object?>(row)).ToList()
            : new List<Dictionary<string, object?>>();

        foreach (var filter in filters)
        {
            rows = rows.Where(row => Matches(row, filter)).ToList();
        }

        if (orderings.Count > 0)
        {
            rows.Sort((left, right) => Compare(left, right, orderings));
        }

        if (offset.HasValue)
        {
            rows = rows.Skip(offset.Value).ToList();
        }

        if (limit.HasValue)
        {
            rows = rows.Take(limit.Value).ToList();
        }

        if (projection.Count > 0)
        {
            rows = rows
                .Select(row => projection.ToDictionary(column => column, column => row.GetValueOrDefault(column)))
                .ToList();
        }

        return rows;
    }

    private static bool Matches(IDictionary<string, object?> row, Filter filter)
    {
        var value = row.GetValueOrDefault(filter.Column);
        return filter.Operator switch
        {
            Operator.Eq => Equals(value, filter.Value),
            Operator.Neq => !Equals(value, filter.Value),
            Operator.In when filter.Value is IEnumerable<object?> collection => collection.Cast<object?>().Contains(value),
            _ => false,
        };
    }

    private static int Compare(IDictionary<string, object?> left, IDictionary<string, object?> right, IReadOnlyList<OrderClause> orderings)
    {
        foreach (var ordering in orderings)
        {
            var comparison = Comparer<object?>.Default.Compare(
                left.GetValueOrDefault(ordering.Column),
                right.GetValueOrDefault(ordering.Column));
            if (comparison == 0)
            {
                continue;
            }

            return ordering.Direction == OrderDirection.Desc ? -comparison : comparison;
        }

        return 0;
    }
}
